dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -7,3263 +7,6479 @@
+  return DDS::RETCODE_UNSUPPORTED;
 }
 
-DDS::ReturnCode_t DynamicDataImpl::get_descriptor(DDS::MemberDescriptor*& value, MemberId id)
+DDS::ReturnCode_t DynamicDataImpl::clear_value(DDS::MemberId /*id*/)
 {
-  const DDS::DynamicType_var base_type = get_base_type(type_);
-  DDS::DynamicTypeMember_var dtm;
-  if (type_->get_member(dtm, id) != DDS::RETCODE_OK) {
-    return DDS::RETCODE_ERROR;
-  }
+  return DDS::RETCODE_UNSUPPORTED;
+}
 
-  return dtm->get_descriptor(value);
+DDS::DynamicData_ptr DynamicDataImpl::loan_value(DDS::MemberId /*id*/)
+{
+  ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: DynamicDataImpl::loan_value: Not implemented\n"));
+  return 0;
 }
 
-DDS::ReturnCode_t DynamicDataImpl::set_descriptor(MemberId, DDS::MemberDescriptor*)
+DDS::ReturnCode_t DynamicDataImpl::return_loaned_value(DDS::DynamicData_ptr /*value*/)
 {
+  ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: DynamicDataImpl::return_loaned_value: Not implemented\n"));
   return DDS::RETCODE_UNSUPPORTED;
 }
 
-MemberId DynamicDataImpl::get_member_id_by_name(const char* name)
+DDS::DynamicData_ptr DynamicDataImpl::clone()
 {
-  const DDS::DynamicType_var base_type = get_base_type(type_);
+  ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: DynamicDataImpl::clone: Not implemented\n"));
+  return 0;
+}
 
-  const TypeKind tk = base_type->get_kind();
-  switch (tk) {
-  case TK_BOOLEAN:
-  case TK_BYTE:
-  case TK_INT16:
-  case TK_INT32:
-  case TK_INT64:
-  case TK_UINT16:
-  case TK_UINT32:
-  case TK_UINT64:
-  case TK_FLOAT32:
-  case TK_FLOAT64:
-  case TK_FLOAT128:
-  case TK_INT8:
-  case TK_UINT8:
-  case TK_CHAR8:
-  case TK_CHAR16:
-  case TK_ENUM:
-    return MEMBER_ID_INVALID;
-  case TK_STRING8:
-  case TK_STRING16:
-  case TK_SEQUENCE:
-  case TK_ARRAY:
-    // Elements of string, sequence, array must be accessed by index.
-    return MEMBER_ID_INVALID;
-  case TK_MAP:
-    // Values in map can be accessed by strings which is converted from map keys.
-    // But need to find out how this conversion works. In the meantime, only allow
-    // accessing map using index.
-    return MEMBER_ID_INVALID;
-  case TK_BITMASK:
-  case TK_STRUCTURE:
-  case TK_UNION:
-    {
-      DDS::DynamicTypeMember_var member;
-      if (base_type->get_member_by_name(member, name) != DDS::RETCODE_OK) {
-        return MEMBER_ID_INVALID;
-      }
-      DDS::MemberDescriptor_var descriptor;
-      if (member->get_descriptor(descriptor) != DDS::RETCODE_OK) {
-        return MEMBER_ID_INVALID;
-      }
-      if (tk == TK_BITMASK) {
-        // Bitmask's flags don't have ID, so use index instead.
-        return descriptor->index();
-      } else {
-        return descriptor->id();
-      }
-    }
+bool DynamicDataImpl::insert_single(DDS::MemberId id, const ACE_OutputCDR::from_int8& value)
+{
+  // The same member might be already written to complex_map_.
+  // Make sure there is only one entry for each member.
+  if (container_.complex_map_.count(id) > 0) {
+    container_.complex_map_.erase(id);
   }
```

Comment:
It's not necessary to check the `count` before `erase`.  Just calling `erase` is equivalent -- if nothing is there for that key then `erase` returns 0 (which is helpful for the `else if` on line 154 below).